### PR TITLE
[libc][math] Implement correctly rounded double precision tan

### LINKED ISSUE #96930: Provide `tan` in libc
We're trying to build Pico SDK against LLVM libc and `tan` is one of the missing functions.

### PR DESCRIPTION
**Pull request by lntue**
Using the same range reduction as `sin`, `cos`, and `sincos`:
1) Reducing `x = k*pi/128 + u`, with `|u| <= pi/256`, and `u` is in double-double.
2) Approximate `tan(u)` using degree-9 Taylor polynomial.
3) Compute
```
   tan(x) ~ (sin(k*pi/128) + tan(u) * cos(k*pi/128)) / (cos(k*pi/128) - tan(u) * sin(k*pi/128))
```
using the fast double-double division algorithm in [the CORE-MATH project](https://gitlab.inria.fr/core-math/core-math/-/blob/master/src/binary64/tan/tan.c#L1855).
4) Perform relative-error Ziv's accuracy test
5) If the accuracy tests failed, we redo the computations using 128-bit precision `DyadicFloat`.

Fixes https://github.com/llvm/llvm-project/issues/96930

### CODE REVIEW

**Comment by lntue** [comment]
@zimmermann6

File: libc/src/__support/FPUtil/double_double.h [hunk 1]
```diff
@@ -129,6 +129,42 @@ LIBC_INLINE DoubleDouble multiply_add<DoubleDouble>(const DoubleDouble &a,
...
+  DoubleDouble r;
+  double q = 1.0 / b.hi;
+  r.hi = a.hi * q;
+
+#ifdef LIBC_TARGET_CPU_HAS_FMA
+  double e_hi = fputil::multiply_add(b.hi, -r.hi, a.hi);
+  double e_lo = fputil::multiply_add(b.lo, -r.hi, a.lo);
+#else
+  DoubleDouble b_hi_r_hi = fputil::exact_mult</*NO_FMA*/ true>(b.hi, -r.hi);
+  DoubleDouble b_lo_r_hi = fputil::exact_mult</*NO_FMA*/ true>(b.lo, -r.hi);
```
Comment (jhuber6) [inline libc/src/__support/FPUtil/double_double.h:159]:
```suggestion
  DoubleDouble b_hi_r_hi = fputil::exact_mult</*NO_FMA=*/true>(b.hi, -r.hi);
  DoubleDouble b_lo_r_hi = fputil::exact_mult</*NO_FMA=*/true>(b.lo, -r.hi);
```
LLVM prefers these comments look like this, don't know if the `libc` style changes it, but `clang-format` should understand it.

File: libc/src/math/generic/CMakeLists.txt [hunk 2]
```diff
@@ -323,6 +323,27 @@ add_entrypoint_object(
...
+    libc.src.errno.errno
+    libc.src.__support.FPUtil.double_double
+    libc.src.__support.FPUtil.dyadic_float
+    libc.src.__support.FPUtil.except_value_utils
+    libc.src.__support.FPUtil.fenv_impl
+    libc.src.__support.FPUtil.fp_bits
+    libc.src.__support.FPUtil.multiply_add
+    libc.src.__support.macros.optimization
+  COMPILE_OPTIONS
+    -O3
```
Comment (jhuber6) [inline libc/src/math/generic/CMakeLists.txt:344]: We don't do debug builds at all for math, right? I think normally a release build is `O2` so it makes sense to override to `O3` in that case.

File: libc/test/src/math/tan_test.cpp [hunk 3]
```diff
@@ -6,27 +6,116 @@
...
 }
+
+TEST_F(LlvmLibcTanTest, InDoubleRange) {
+  constexpr uint64_t COUNT = 1'234'51;
+  uint64_t START = LIBC_NAMESPACE::fputil::FPBits<double>(0x1.0p-50).uintval();
+  uint64_t STOP = LIBC_NAMESPACE::fputil::FPBits<double>(0x1.0p200).uintval();
+  uint64_t STEP = (STOP - START) / COUNT;
+
+  auto test = [&](mpfr::RoundingMode rounding_mode) {
+    mpfr::ForceRoundingMode __r(rounding_mode);
```
Comment (jhuber6) [inline libc/test/src/math/tan_test.cpp:64]: Aren't double underscore names all reserved by the implementation? I'd say we *are* the implementation, but these tests can be run in overlay mode, right?

**Review by jhuber6 — APPROVED** [review]
I can't vouch for the math implementation, but the patch looks good with a handful of nits.

File: libc/test/src/math/tan_test.cpp (hunk 3 above)
Comment (lntue) [reply libc/test/src/math/tan_test.cpp]: Updated the name.

File: libc/src/__support/FPUtil/double_double.h (hunk 1 above)
Comment (lntue) [reply libc/src/__support/FPUtil/double_double.h]: Done.

**Review by jhuber6 — APPROVED** [review]

**Pull request merged**